### PR TITLE
Upgrading to tensorflow==2.7.4 + work in 2022

### PR DESCRIPTION
**Pull request by lastmansleeping**

### CODE REVIEW

File: python/ml4ir/applications/classification/tests/test_classification_model.py [hunk 1]
```diff
@@ -42,7 +42,7 @@ def test_group_metrics_df(self):
         Test the dimensions of the grouped metrics
         """
         # Metrics cardinality and names
-        metrics = self.classification_pipeline.metrics_keys  # Metrics during training
+        metrics = ["loss"] + self.classification_pipeline.metrics_keys  # Metrics during training
```
Comment (ArmandGiraud) [inline python/ml4ir/applications/classification/tests/test_classification_model.py:45]: I'm not certain where this is defined,but maybe self.classification_pipeline.metrics_keys.insert(0, "loss")

File: python/ml4ir/applications/ranking/tests/test_base.py [hunk 2]
```diff
@@ -41,7 +41,7 @@ def get_ranking_model(
         loss_key: str,
         metrics_keys: List,
         feature_config: FeatureConfig,
-        model_config: dict = {},
+        model_config: dict = None,
```
Comment (ArmandGiraud) [inline python/ml4ir/applications/ranking/tests/test_base.py:44]: well, that looks very suspicious.., at least remove type hints?

File: python/ml4ir/applications/ranking/tests/test_base.py [hunk 3]
```diff
@@ -51,6 +51,7 @@ def get_ranking_model(
 
         NOTE: Override this method to create custom loss, scorer, model objects
         """
+        self.model_config = model_config if model_config else self.model_config
```
Comment (ArmandGiraud) [inline python/ml4ir/applications/ranking/tests/test_base.py:54]: Empty dict evaluates to false with bool({}).

File: python/ml4ir/applications/ranking/tests/test_transfer_learning.py [hunk 4]
```diff
@@ -72,6 +73,8 @@ def get_ranking_dataset_and_model(
 
         return ranking_model, relevance_dataset
 
+    @unittest.skip("""Disabled as transfer learning does not work currently.
```
Comment (ArmandGiraud) [inline python/ml4ir/applications/ranking/tests/test_transfer_learning.py:76]: idem

File: python/ml4ir/base/config/keys.py [hunk 5]
```diff
@@ -21,6 +21,7 @@ class ArchitectureKey(Key):
     DNN = "dnn"
```
Comment (ArmandGiraud) [inline python/ml4ir/base/config/keys.py:21]: Not related, but is there a reason for not using enums here?

File: python/ml4ir/base/config/parse_args.py [hunk 6]
```diff
@@ -136,6 +129,15 @@ def define_args(self):
             help="A space separated list of metrics to compute.",
         )
 
+        self.add_argument(
```
Comment (ArmandGiraud) [inline python/ml4ir/base/config/parse_args.py:132]: Not directly related, but I counted 46 add_argument in this file, should we consider moving this arguments to a conf file at some points?

File: python/ml4ir/base/data/tfrecord_reader.py [hunk 7]
```diff
@@ -261,13 +261,8 @@ def _parse_fn(proto):
...
 
-            if self.feature_config.aux_label:
-                aux_labels = features_dict.get(self.feature_config.get_aux_label(key="node_name"))
-                # return X and y and y_aux which can be used with fit(), predict() and evaluate()
-                return features_dict, {self.output_name: labels, self.aux_output_name: aux_labels}
-            else:
-                # return X and y which can be used with fit(), predict() and evaluate()
-                return features_dict, labels
+            # return X and y which can be used with fit(), predict() and evaluate()
+            return features_dict, labels
```
Comment (ArmandGiraud) [inline python/ml4ir/base/data/tfrecord_reader.py:265]: Not sure what's going on with auxiliary loss here, are they entirely removed, renamed, or refactored?

File: python/ml4ir/base/io/local_io.py [hunk 8]
```diff
@@ -45,7 +45,7 @@ def make_directory(self, dir_path: str, clear_dir: bool = False):
         os.makedirs(dir_path)
 
     def read_df(
-        self, infile: str, sep: str = ",", index_col: int = None
+        self, infile: str, sep: str = ",", index_col: int = None, **kwargs
```
Comment (ArmandGiraud) [inline python/ml4ir/base/io/local_io.py:48]: what's the reason for using kwargs instead of argument here?

File: python/ml4ir/base/model/architectures/auto_dag_network.py [hunk 9]
```diff
@@ -0,0 +1,448 @@
+from __future__ import annotations
+
```
Comment (ArmandGiraud) [inline python/ml4ir/base/model/architectures/auto_dag_network.py:2]: Should definitely not be included in this PR IMHO

File: python/ml4ir/base/model/layers/tf_native.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
+import tensorflow as tf
+from tensorflow.keras import layers
+
+
+class TFNativeLayer(layers.Layer):
```
Comment (ArmandGiraud) [inline python/ml4ir/base/model/layers/tf_native.py:5]: where is this used?

File: python/ml4ir/base/model/losses/loss_base.py [hunk 11]
```diff
@@ -1,39 +1,56 @@
...
 
         Returns
         -------
-        function
-            Final activation function that is applied on the scores
-            before computing losses
+        tensor
+            Resulting score tensor after applying the function on the logits
         """
         raise NotImplementedError
```
Comment (ArmandGiraud) [inline python/ml4ir/base/model/losses/loss_base.py:48]: That's a lot of doc for this function :D

File: python/ml4ir/base/tests/test_lr_schedule.py [hunk 12]
```diff
@@ -217,7 +218,7 @@ def test_cyclic_lr_in_training_pipeline(self):
         lr_list = my_callback_object.get_lr_list()
         lr_gold = [0.001,0.020800006,0.040599994,0.0604,0.080199994,0.1,0.080199994,0.0604,0.040599994,0.020800006,0.001,
                    0.010900003,0.020800006,0.030699994,0.040599994,0.050499998,0.040599994,0.030699994,0.020800006,
-                   0.010900003,0.001,0.0059499955,0.010900003,0.015849996,0.020800006,0.02575,0.020800006,0.015849996,
+                   0.010900003,0.001,0.0059499955,0.010900003,0.015849996,0.020800006,0.02575,0.02575,0.020800006,0.015849996,
```
Comment (ArmandGiraud) [inline python/ml4ir/base/tests/test_lr_schedule.py:221]: Well, not entirely clear what's going on here

File: python/ml4ir/build/Dockerfile [hunk 13]
```diff
@@ -9,6 +9,9 @@ ENV PYTHONPATH=/home/ml4ir/python
 
 # Install python dependencies
 COPY requirements.txt /home/ml4ir/python/
+RUN sudo apt-get --allow-releaseinfo-change update
+# Note: This is needed for DAG visualization
```
Comment (ArmandGiraud) [inline python/ml4ir/build/Dockerfile:13]: The real question is why do we need to visualize DAGs? how more informative is it from a print?

**Review by ArmandGiraud — COMMENTED** [review]
The general feedback on this PR is, do not merge it unless you are very confident that those changes will not yield any future problems, and that the code coverage is very good.  

Given the complex nature of the project, mixing java and python, and having a child repo
my recommandation is to only keep what is necessary to upgrade tensorflow. The rest could be added in following PRs. 
It will help check the different changes more carefull  add more context, and proceed step by step.

File: python/ml4ir/build/Dockerfile (hunk 13 above)
Comment (lastmansleeping) [reply python/ml4ir/build/Dockerfile:13]: These are from PRs already merged to `dev_tf27` while we were working on semantic search and more complex architectures.

**Comment by lastmansleeping** [comment]
> Given the complex nature of the project, mixing java and python, and having a child repo
> my recommandation is to only keep what is necessary to upgrade tensorflow. The rest could be added in following PRs.
> It will help check the different changes more carefull add more context, and proceed step by step.

I am not sure how to create a PR with only the tensorflow upgrade as we have been using this branch to work on the tf27 version for about a year now. All the commits into this branch have been peer reviewed PRs with test coverage. Happy to hear any ideas.

**Comment by ArmandGiraud** [comment]
@lastmansleeping closing this one
